Refactor profile connection flow to async per-device manager

zhongzhijie1 opened the pull request:
- Add profile connection entry type and bt_profile_conn_entry_t structure.
  - Introduce bt_sal_profile_connect_request, bt_sal_profile_disconnect_register and bt_sal_profile_disconnect APIs.
  - Add internal worker APIs: bt_sal_remove_bond_internal, bt_sal_disconnect_internal.
  - Expand connection manager data model:
    - separate connecting / disconnecting manager lists
    - support dynamic profile_conn_entries per device
    - change profile flags to 64-bit and add profile_id -> flag helper
    - add async request wrapper and service-worker invocation for profile actions
  - Implement profile connect/disconnect orchestration:
    - trigger profile actions on ACL connect
    - ensure profile disconnect handlers run before ACL teardown
    - handle unpair flow by removing bond after ACL disconnect
  - Update A2DP/AVRCP and adapter code to use new APIs:
    - register profile disconnect handlers and notify profile-connected events
    - use bt_sal_profile_connect_request / bt_sal_profile_disconnect wrappers
    - invoke cm callbacks on ACL connect/disconnect paths

chengkai15 commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -22,8 +22,21 @@ typedef struct {
     uint8_t profile_id;
 } cm_data_t;
 
+typedef bt_status_t (*bt_profile_connect_entry_func_t)(bt_controller_id_t id, bt_address_t* addr);
```
please break into smaller  reviewable commits

zhongzhijie1 replied:
done

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -60,6 +60,18 @@ static inline uint64_t profile_id_to_flag(uint8_t profile_id)
+        return false;
+    }
+
+    return entry->func == *target_func;
```
could we record func ptr?

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -146,6 +158,43 @@ void cm_data_destory(cm_data_t* data)
+    if (!bt_sal_connecting_list || !list) {
+        return BT_STATUS_NOMEM;
+    }
+
```
if profile_flags was invalid, the  return Immediately? and we do not need to  check it with every profile

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -146,6 +158,43 @@ void cm_data_destory(cm_data_t* data)
+            continue;
+        }
+
+        /* TODO: async invoke to service_worker thread */
```
when it would invork func run?

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -306,6 +355,70 @@ bt_status_t bt_sal_cm_try_disconnect_profiles(bt_address_t* addr, bool is_unpair
+    return BT_STATUS_SUCCESS;
+}
+
+bt_status_t bt_sal_request_profile_connect(bt_address_t* addr)
```
could we remove bt_sal_request_profile_connect?

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -189,7 +285,17 @@ static bt_status_t bt_sal_trigger_profile_connects(bt_profile_connection_manager
+            if (bt_list_is_empty(manager->profile_connect_entries)) {
+                bt_list_remove(bt_sal_connecting_list, manager);
+            }
+            continue;
```
remove  continue

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -338,41 +338,94 @@ static void bt_sal_cm_profile_disconnected(void* data)
-        break;
+
+    if (bt_sal_disconnecting_list != NULL) {
+        manager = bt_list_find(bt_sal_disconnecting_list, bt_connection_manager_find, &cm_data->addr);
```
remove duplicate codes  with manager?

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -338,41 +338,94 @@ static void bt_sal_cm_profile_disconnected(void* data)
+    if (bt_sal_disconnecting_list != NULL) {
+        manager = bt_list_find(bt_sal_disconnecting_list, bt_connection_manager_find, &cm_data->addr);
+
+        if (manager != NULL) {
```
could we consturct these codes with one line. like

flags_clear(&manager->profile_flags, profile)

chengkai15 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -385,23 +438,35 @@ static void bt_sal_cm_acl_disconnected(void* data)
-        BT_LOGW("%s, manager not found.", __func__);
-        cm_data_destory(cm_data);
-        return;
+    if (bt_sal_disconnecting_list != NULL) {
```
duplicated codes with LINE:441 to LINE:447?

chengkai15 commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -1576,14 +1590,28 @@ bt_status_t bt_sal_a2dp_source_connect(bt_controller_id_t id, bt_address_t* addr
+        return status;
+    }
+
+    return bt_sal_request_profile_connect(addr);
```
remove

zhongzhijie1 committed:
add direct profile connect/disconnect flow and API docs

bug: v/77661

- Add profile connection handler type and bt_profile_conn_handler_node_t structure.
- Introduce bt_sal_profile_connect_request, bt_sal_profile_disconnect_register and bt_sal_profile_disconnect APIs.
- Add internal worker APIs: bt_sal_remove_bond_internal, bt_sal_disconnect_internal.
- Expand connection manager data model:
  - separate connecting / disconnecting manager lists
  - support dynamic profile_conn_handler_node per device
  - change profile flags to 64-bit and add profile_id -> flag helper
  - add async request wrapper and service-worker invocation for profile actions
- Implement profile connect/disconnect orchestration:
  - trigger profile actions on ACL connect
  - ensure profile disconnect handlers run before ACL teardown
  - handle unpair flow by removing bond after ACL disconnect
- Update A2DP/AVRCP and adapter code to use new APIs:
  - register profile disconnect handlers and notify profile-connected events
  - use bt_sal_profile_connect_request / bt_sal_profile_disconnect wrappers
  - invoke cm callbacks on ACL connect/disconnect paths

Signed-off-by: zhongzhijie1 <zhongzhijie1@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
-    bt_address_t addr;
-    uint8_t profile_id;
+    bt_address_t addr; /**< Remote device address */
+    uint8_t profile_id; /**< Profile identifier (PROFILE_*) */
```
Add an actual example rather than `PROFILE_`

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * profile operations should be executed for a specific device.
+ */
+typedef struct {
+    bt_profile_conn_entry_func_t func; /**< Profile action function */
```
`func` might be a `cb`

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ *     // Establish profile-specific session
+ *     return BT_STATUS_SUCCESS;
+ * }
+ *
```
Remove examples when it's too simple (almost transparent)

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * @brief Register a profile connect request for a remote device.
+ *
+ * Registers a profile-connect entry for the given remote device.
+ * The provided callback function is executed in service-worker context to
```
`executed` -> `invoked`

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+/**
+ * @brief Internal ACL disconnect request executed in worker context.
+ */
+bt_status_t bt_sal_disconnect_internal(bt_controller_id_t id,
```
Do we need a `transport` as well?

zhongzhijie1 replied:
no need, only for classic

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+/**
+ * @brief Enqueue a profile-connected event.
+ */
+void bt_sal_cm_profile_connected_callback(cm_data_t* data);
```
Can we move `cm_data_new` inside this function?
So we no longer need to expose `cm_data_new`

zhongzhijie1 replied:
current not recomment

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
-bool bt_sal_avrcp_try_disconnect_avrcp_control(bt_controller_id_t id, bt_address_t* addr);
\ No newline at end of file
+/* -------------------------------------------------------------------------- */
+/*  Profile Disconnect & ACL Teardown Orchestration                           */
```
Is `Orchestration` often used?

gzh-terry replied:
Try to simplify these documentation.

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ *   it triggers ACL disconnect immediately.
+ * - Otherwise, it triggers the corresponding profile-level disconnect actions.
+ *
+ * ### Internally, this function performs:
```
Remove internal descriptions.

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
 } cm_data_t;
 
+/**
+ * @brief Profile connect/disconnect entry function type.
```
This is literally not an "entry".

One can be a "entry" function when a groups of functions are start from one point (for example, a `main()`, or a `a2dp_service_interface()`). However, the profile connection procedure are not start from the `bt_profile_conn_entry_func_t`.

I would recommend to call it a `callback` (as how it actually performs).

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ *
+ * This function type is used to invoke profile-specific connect or disconnect
+ * operations. Implementations should perform the profile action synchronously
+ * (in the service worker context) and return a bt_status_t indicating result.
```
Why it should performed synchronously? Can I post a message and do the connection elsewhere?

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * (in the service worker context) and return a bt_status_t indicating result.
+ *
+ * @param id    Controller/adapter ID.
+ * @param addr  Pointer to remote device address.
```
Pointer to `the` remote device address.

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+typedef struct {
+    bt_profile_conn_entry_func_t func; /**< Profile action function */
+    uint8_t profile_id; /**< Profile id (PROFILE_*) */
+    bt_controller_id_t id; /**< Controller/adapter id */
```
Can I have one registration for ALL the adapters?

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * perform the actual profile-level connection procedure.
+ *
+ * Typical usage:
+ * 1. Profile module implements <profile>_profile_connect() as its connect handler.
```
Try some other descriptions, `<profile>_profile_connect()` is hard to understand.

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * @brief Register a profile disconnect handler for a remote device.
+ *
+ * Registers a profile-disconnect entry for the device.
+ * When the connection manager decides to disconnect ACL,
```
I'm not quite sure, is this a part of our requirement?
Probably the "service connection manager" should do this?

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * }
+ *
+ * // Example: register disconnect handler when profile becomes connected
+ * static void on_a2dp_sink_connected(...)
```
Remove the sample. It's too long and not really informative.

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
 typedef struct {
-    bt_address_t addr;
-    uint8_t profile_id;
+    bt_address_t addr; /**< Remote device address */
```
Do we add `address_type`?

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Initiate profile-level disconnect for a device before ACL teardown.
```
`disconnect` -> `disconnection`

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ *
+ *   1. All registered profile disconnect handlers for the device are executed
+ *      first (A2DP, HFP, PBAP, …).
+ *   2. Only after all profile disconnects finish (or no profiles exist),
```
`disconnect` is a verb (in most cases).
So we can say `after all profiles are disconnected` or `after all profile disconnections are finished`.

Although it is not often used, but we can also have `disconnect` as a none. Then we can say `after all profile disconnects`

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ *   2. Only after all profile disconnects finish (or no profiles exist),
+ *      the ACL link will be disconnected.
+ *   3. If @p is_unpair is true, bond information will be removed after ACL
+ *      teardown.
```
I though bond information will be remove before (or during) ACL teardown.

Suppose the bond information is stored in ACL, then it is removed by `bt_br_unpair`, which actually happens before ACL disconnection.

zhongzhijie1 replied:
key info saves in bt_dev structure, graceful terminate acl, then bond info

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ *   3. If @p is_unpair is true, bond information will be removed after ACL
+ *      teardown.
+ *
+ * The function is designed to be **idempotent** and **progressive**:
```
Try to simplify the documentations, and use easy words.

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -17,19 +17,310 @@
+ * @see bt_sal_remove_bond_internal
+ * @see bt_sal_cm_acl_disconnected_callback
+ */
+bt_status_t bt_sal_cm_try_disconnect_profiles(bt_address_t* addr, bool is_unpair);
```
It not only try to disconnect profiles, but also try to disconnect the ACL.

BTW, do we need a `transport` here?

zhongzhijie1 replied:
currently not , only for classic bluetooth

gzh-terry commented on service/stacks/zephyr/sal_adapter_interface.c:
```diff
@@ -1394,19 +1422,39 @@ bt_status_t bt_sal_cancel_bond(bt_controller_id_t id, bt_address_t* addr, bt_tra
+    if (!req)
+        return BT_STATUS_NOMEM;
+
+    status = bt_sal_cm_try_disconnect_profiles(addr, true);
```
`bt_sal_cm_try_disconnect_profiles` first.
`sal_adapter_req` later.

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -65,6 +111,127 @@ static bool bt_cm_disconnect_find(void* data, void* context)
+        return;
+    }
+
+    entry = bt_list_find(manager->profile_conn_entries, match_profile_id, (void*)&req->profile_id);
```
I almost forget why do we need this `profile_id`?

zhongzhijie1 replied:
Consider a scenario where profile_id is associated with profile_flags to prevent duplicate requests. And it also plays a role in matching the service work requirements with the actually registered handler_node. In other words, it is not completely ineffective. Currently, it is not recommended to remove it.

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -226,30 +529,133 @@ bt_status_t bt_sal_cm_try_disconnect_profiles(bt_address_t* addr, bool is_unpair
+    }
+
+    entry_node->func = func;
+    entry_node->profile_id = profile_id;
```
Is it possible that we only record the `func`, but not the `profile_id`.
It looks like nobody cares about the `profile_id`.

zhongzhijie1 replied:
Consider a scenario where profile_id is associated with profile_flags to prevent duplicate requests. And it also plays a role in matching the service work requirements with the actually registered handler_node. In other words, it is not completely ineffective. Currently, it is not recommended to remove it.

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -65,6 +111,127 @@ static bool bt_cm_disconnect_find(void* data, void* context)
+        return;
+    }
+
+    status = req->func(req->id, &req->device_addr);
```
Should we check `manager->profile_flags` before calling `req->func` to avoid duplicated connections?

zhongzhijie1 replied:
we have checked profile_flags before submit this work request, so no need check here

gzh-terry replied:
It doesn't make sense.

We check flags when submit work, and set flags when the work is (almost) done.

If the upper layer send the request twice consecutively. It would pass the check anyway because any of the works are not started yet.

gzh-terry replied:
> It doesn't make sense.
> 
> We check flags when submit work, and set flags when the work is (almost) done.
> 
> If the upper layer send the request twice consecutively. It would pass the check anyway because any of the works are not started yet.

Example:

* App::connection_request -> check flag (pass)
* App::connection_request -> check flag (pass)

(after bluetoothd is idle)
* Work::invoke -> set flag
* Work::invoke -> set flag

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -81,41 +248,75 @@ cm_data_t* cm_data_new(bt_address_t* addr, uint8_t profile_id)
 
     if (manager->profile_flags != FLAG_NONE) {
-        BT_LOGI("%s, Disconnecting profile.", __func__);
+        BT_LOGW("%s, Disconnecting profile.", __func__);
```
This is not a warning.

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -81,41 +248,75 @@ cm_data_t* cm_data_new(bt_address_t* addr, uint8_t profile_id)
         BT_LOGE("%s, conn not found.", __func__);
         return BT_STATUS_FAIL;
+    } else {
+        return bt_sal_disconnect_internal(PRIMARY_ADAPTER, &manager->device_addr, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
```
Remember to `bt_conn_unref(conn)`, which is acquired from `bt_conn_lookup_addr_br` at line 313.

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -81,41 +248,75 @@ cm_data_t* cm_data_new(bt_address_t* addr, uint8_t profile_id)
     if (conn == NULL) {
         BT_LOGE("%s, conn not found.", __func__);
         return BT_STATUS_FAIL;
+    } else {
```
Redundant `else` since we already have a `return` in the previous `if`

gzh-terry replied:
BTW, we can actually check `conn` inside `STACK_CALL(disconnect)` rather than here.
It makes no sense that we check `conn` now but use it later in a worker thread.

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -125,41 +326,86 @@ static void bt_sal_cm_profile_disconnected(void* data)
+
+                if (bt_list_is_empty(manager->profile_conn_entries)) {
+                    bt_try_disconnect_acl(manager);
+                    bt_list_remove(bt_sal_connecting_list, manager);
```
`bt_sal_connecting_list` -> `bt_sal_disconnecting_list `

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -125,41 +326,86 @@ static void bt_sal_cm_profile_disconnected(void* data)
+                    bt_list_remove(bt_sal_connecting_list, manager);
+                }
+            } else {
+                BT_LOGW("%s, entry not found.", __func__);
```
Can we extract these into a `void remove_from_connection_manager_list(*list, profile_id)`?

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -172,23 +418,39 @@ static void bt_sal_cm_acl_disconnected(void* data)
+    if (bt_sal_connecting_list != NULL) {
+        manager = bt_list_find(bt_sal_connecting_list, bt_connection_manager_find, &cm_data->addr);
+        if (manager != NULL) {
+            bt_list_remove(bt_sal_disconnecting_list, manager);
```
`bt_sal_disconnecting_list` -> `bt_sal_connecting_list`

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -208,15 +478,48 @@ void bt_sal_cm_acl_disconnected_callback(cm_data_t* data)
         return BT_STATUS_FAIL;
 
-    manager = bt_list_find(bt_sal_disconnecting_list, bt_cm_disconnect_find, addr);
+    conn = bt_conn_lookup_addr_br((bt_addr_t*)addr);
```
Remember to `bt_conn_unref(conn)` for each `conn = bt_conn_lookup_addr_br`

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -208,15 +478,48 @@ void bt_sal_cm_acl_disconnected_callback(cm_data_t* data)
+            return BT_STATUS_SUCCESS;
+        }
+
+        manager->is_unpair = manager->is_unpair || is_unpair;
```
It is just checked `manager->is_unpair` can not be `true`
So we can have `manager->is_unpair = is_unpair` now.

zhongzhijie1 replied:
if already is_unpair, skip, if new request is is_unpair = true, such as user call disconnect, then call removebond imediately, or first call removebond, then call disconnect.

this code can handle it

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -226,30 +529,133 @@ bt_status_t bt_sal_cm_try_disconnect_profiles(bt_address_t* addr, bool is_unpair
+    if (!manager)
+        return BT_STATUS_NOMEM;
+
+    conn = bt_conn_lookup_addr_br((bt_addr_t*)addr);
```
Remember `bt_conn_unref`

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/include/sal_connection_manager.h:
```diff
@@ -22,14 +22,35 @@ typedef struct {
+bt_status_t bt_sal_profile_disconnect_register(bt_address_t* addr,
+    uint8_t profile_id, bt_controller_id_t id,
+    bt_profile_conn_handler_t handler);
+bt_status_t bt_sal_profile_disconnect(bt_address_t* addr,
```
`bt_sal_profile_disconnect` -> `bt_sal_profile_disconnect_request`

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -30,33 +30,79 @@
-    *profile_flags = FLAG_NONE;
+    SAL_ASSERT(profile_id <= PROFILE_MAX);
+
+    return (1UL << (profile_id));
```
`1ULL` for `uint64_t`

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -208,15 +469,50 @@ void bt_sal_cm_acl_disconnected_callback(cm_data_t* data)
+        return BT_STATUS_FAIL;
+    }
+
+    bt_conn_unref(conn);
```
`bt_conn_unref` after `bt_conn_get_info`

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -226,30 +522,135 @@ bt_status_t bt_sal_cm_try_disconnect_profiles(bt_address_t* addr, bool is_unpair
+        return bt_sal_connect(PRIMARY_ADAPTER, addr);
+    }
+
+    bt_conn_unref(conn);
```
`bt_conn_unref` after `bt_conn_get_info`

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -65,6 +111,126 @@ static bool bt_cm_disconnect_find(void* data, void* context)
+
+    SAL_ASSERT(req);
+
+    if (!req->user_data) {
```
Add comment here: `!req->user_data` means a "disconnection", otherwise it's a "connection"

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -65,6 +111,126 @@ static bool bt_cm_disconnect_find(void* data, void* context)
+    if (!bt_list_find(manager->profile_conn_handler_list, match_profile_func, (void*)&req->handler)) {
+        BT_LOGW("%s, handler_node handler not found.", __func__);
+        free(req);
+        return;
```
Clear `manager->profile_flags` on this return.

zhongzhijie1 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -65,6 +111,126 @@ static bool bt_cm_disconnect_find(void* data, void* context)
+    if (!handler_node) {
+        BT_LOGW("%s, handler_node not found.", __func__);
+        free(req);
+        return;
```
Clear `manager->profile_flags` on this return.

zhongzhijie1 replied:
done

liuX10 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -23,40 +23,86 @@
+static bool match_profile_func(void* data, void* context)
+{
+    bt_profile_conn_handler_node_t* handler_node = (bt_profile_conn_handler_node_t*)data;
+    bt_profile_conn_handler_t* target_func = (bt_profile_conn_handler_t*)context;
```
why use `data `and `context `two params for a same address.
```
typedef struct {
    bt_profile_conn_handler_t handler;
    uint8_t profile_id;
    bt_controller_id_t id;
} bt_profile_conn_handler_node_t;
```

zhongzhijie1 replied:
data is from lists, context is input func_handler parameter
I want find one func_handler through lists.

liuX10 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -23,40 +23,86 @@
+    return handler_node->handler == *target_func;
+}
+
+static bool match_profile_id(void* data, void* context)
```
if context only for profile_id, can directly use profile_id?.

zhongzhijie1 replied:
we use profile_id to find one bt_profile_conn_handler_node_t, use bt_profile_conn_handler_t handler

liuX10 commented on service/stacks/zephyr/sal_connection_manager.c:
```diff
@@ -65,6 +111,126 @@ static bool bt_cm_disconnect_find(void* data, void* context)
     return memcmp(&manager->device_addr, context, sizeof(bt_address_t)) == 0;
 }
 
+static void profile_entry_destroy(void* data)
+{
+    if (data) {
+        bt_profile_conn_handler_node_t* handler_node = (bt_profile_conn_handler_node_t*)data;
```
I see this func just free a variable, is necessary？

zhongzhijie1 replied:
it is useful for 
`manager->profile_conn_handler_list = bt_list_new(profile_entry_destroy);`

when 

`bt_list_remove(manager->profile_conn_handler_list, handler_node);`
